Add NB for BERT models

hkirvesl opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the guidelines for contributing: https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines
-->

**Reference issues/PRs**

Fixes #81 By adding a notebook on BERT models with topological interpretability methods built on top of one. The old tutorials on Q&A and Translation remain for now.


**Types of changes**
<!--
What types of changes does your code introduce? Put an `x` in all the boxes that apply.
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Description**

1) Add notebook for training HuggingFace Bert model with giotto-deep, and how to use topological pruning methods for such models.
2) Add support for training HuggingFace models with an unwrapper in Trainer.py
3) Modify models/utils.py to support the the HuggingFace models (now with an unwrapper)

**Screenshots (if appropriate)**

**Any other comments?**

**Checklist**
<!--
Go over all the following points, and put an `x` in all the boxes that apply.
If you're unsure about any of these, don't hesitate to ask. We're here to help!
-->
- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed. I used `pytest` to check this on Python tests.

<!--
We value all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Thanks for contributing!
-->

matteocao commented on examples/basic_tutorial_BERT.ipynb:
```diff
@@ -0,0 +1,1172 @@
+  "kernelspec": {
+   "display_name": "Python [conda env:torch] *",
+   "language": "python",
+   "name": "conda-env-torch-py"
```
the CI is complaining about this name: 

```
jupyter_client.kernelspec.NoSuchKernel: No such kernel named conda-env-torch-py
```

Can you please change this to:
```
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
```

hkirvesl replied:
Thank you! Should be fixed now

matteocao replied:
@hkirvesl can you please use `os.path.join` to build file path please? windows and posix OS have a different convention (back and forward slashes)

<img width="1205" alt="Screen Shot 2022-09-29 at 19 56 54" src="https://user-images.githubusercontent.com/38946719/193107037-1447a39c-3506-4e18-baef-2a8131209c19.png">

hkirvesl replied:
Thank you for the suggestion. Done.

matteocao reviewed (changes requested):
Once the CI pass, I will study your notebook in details, as I am really curious about it!

matteocao commented on gdeep/trainer/trainer.py:
```diff
@@ -266,9 +266,13 @@ def _send_to_device(
             x = new_x
             prediction = self.model(*x)
+            if(hasattr(prediction, "logits")): # unwrapper for HuggingFace BERT model
+                prediction = prediction.logits # unwrapper for HuggingFace BERT model
```
all good here!

matteocao reviewed (changes requested):
Dear @hkirvesl ,

I think we are almost there: I only have a few minor comments.
I will then review the notebbook in detail.

**One important note:** did you pull master to your fork before committing? I ask because there is one item in the CI that took > 5h to run and has currently not yet finished (though it should really take 25 mins at most) and I wonder if it is related to not having integrated the fixes. If you have done so, than ignore this last comment

raphaelreinauer commented on .pre-commit-config.yaml:
```diff
@@ -6,11 +6,11 @@ repos:
+#-   repo: https://github.com/psf/black
+#    rev: 22.8.0
+#    hooks:
+#    -   id: black
```
Could you please wait until my pull request is accepted and then pull again from master? I fixed all mypy errors, and the pre-commit does work now.

hkirvesl replied:
Thanks for the comments, Matteo and Raphael. I did recall some discussion about problems with the prehooks and commenting them out being the recommended temporary work around. I just forgot the details and ended up commenting out too much. All these should be fixed now. Thank you Raphael for the fix, this makes our lives much easier.

matteocao commented on examples/basic_tutorial_BERT.ipynb:
```diff
@@ -0,0 +1,1198 @@
+{
```
<p>here you can simply use</p><pre class="ql-syntax" spellcheck="false">from gdeep.utility import DEVICE
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

hkirvesl replied:
Thanks! It is a more giotto-way to do this. In general, any other comments on how to make the notebook more giotto-like are  highly appreciated.

matteocao commented on examples/basic_tutorial_BERT.ipynb:
```diff
@@ -0,0 +1,1198 @@
+{
```
<p>ca you please preceed this cell with a markdown cell describing what you are doing here?</p><p> </p><p>if this cell and the above one are related to handling masks, then maybe it is worth putting a bit of explanation 2 cells above, details the steps that one needs to take</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

hkirvesl replied:
<p>Good idea.I added a brief note (and deleted an unnecessary chunk)</p>

matteocao commented on examples/basic_tutorial_BERT.ipynb:
```diff
@@ -0,0 +1,1198 @@
+{
```
<p>can you please add a few inline comments to describe what you are doing here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

matteocao replied:
<p>some more comments here please</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

matteocao replied:
<p>if these two cells are "almost" duplicated, can you please create a function in one cell and call it with different parameters?</p><p> </p><p>Probably same comments may work a few cells above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

matteocao replied:
<p>maybe one sentence describe the two cells below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

matteocao commented on examples/basic_tutorial_BERT.ipynb:
```diff
@@ -0,0 +1,1198 @@
+{
```
<p>the comment here , I think, needs to be corrected: <code>raw</code> -&gt; <code>mean aggregated</code>   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

hkirvesl replied:
<p>You are absolutely right. Fixed.</p>

matteocao commented on examples/basic_tutorial_BERT.ipynb:
```diff
@@ -0,0 +1,1198 @@
+{
```
<p>typo: At such --&gt; As such</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

hkirvesl replied:
<p>Good catch!</p>

matteocao commented on examples/basic_tutorial_BERT.ipynb:
```diff
@@ -0,0 +1,1198 @@
+{
```
<p>ca you clarify why you compute the gradients (and of what) in a sentence in a .md cell above this one?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/giotto-ai/giotto-deep/pull/114/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

hkirvesl replied:
<p>Added some discussion on this. Also: If you know of a more giotto-like way of getting your hands on these gradients I think it would improve this notebook.</p>

hkirvesl commented:
Thanks for the good comments! I pulled the main and it did fix a lot, (and also managed to close this PR.) Will reopen once I've cleaned the notebook more a bit more as per the suggestions

raphaelreinauer commented:
Could you please install pre-commit (see README.md)

hkirvesl commented:
Reopening the PR

Changelog:

**The notebook:**

1) Imports

1.1) Removed unneeded imports:
from gdeep.visualisation import persistence_diagrams_of_activations
from gdeep.visualisation import Visualiser

1.2) Added import

from gdeep.utility import DEVICE

2) Cleaned some variable names

3) Added more explanation as per the suggested comments

4) Removed unnecessary code chunks

5) renamed variables

6) Added functions to reduce redundant code

7) Add legend to a persistence diagram plot.

**Testing:**

1) Run all the prehooks, both black and mypy

**models/utils.py**

1) Replace the modification to retain the existing functionality

hkirvesl commented on gdeep/models/utils.py:
```diff
@@ -70,6 +70,6 @@ def __call__(
         """
         if isinstance(module_out, tuple):
-            self.outputs.append(module_out[0].detach())
+            self.outputs.append(module_out[0])#.detach())
```
Thanks for the comment, this is a very good point indeed. I do understand the consequences and I agree what was proposed is not a good solution. To be honest, I forgot about this myself. It was just a quick and dirty fix to make the BERT models work and since it worked and seemingly didn't break any tests I just proceeded without worrying about it.

A good solution would keep the old functionality where applicable, in the same spirit as the change in trainer.py. I believe this has now been fixed.

matteocao reviewed (approved):
this loks good and it is very interesting btw! thanks @hkirvesl !!
